### ISSUE TITLE
Report generation (-p) does not work well for subworkflows

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SoS#979]
Currently reports are generated from signature file of the master workflows... We should either combined all workflow signatures or read all workflow signatures.

**Comment by gaow** [comment]
Sorry for being off-grid -- busy recovering from sickness and cramming for ASHG abstract. ... Just checked out the new report feature. We are not yet reporting CPU / memory etc, right?

**Comment by BoPeng** [comment]
It does (as of this morning) but does not support subworkflow (addressing now).

**Comment by BoPeng** [comment]
Just submitted. Updating another ticket for details.

**Comment by BoPeng** [comment]
We actually have the detailed information on each input and output file (size etc) so we could introduce some method to expand the input output ...

**Comment by BoPeng** [comment]
More stuff to add to the report?

1. Work directory (project directory)
2. Path of the script (does it matter?)
3. sos version
4. workflow ID.
5. A footer with sos link etc.

Nextflow has a timeline plot, which seems to be interesting.

**Comment by BoPeng** [comment]
Regarding the time plot, we might be able to do something like this:

1. get the total duration from the master workflow (end - start).
2. for each task, step, workflow etc when we have start time and duration, we can calculate three percentages, namely percent before start, percent for work and percent after work.
3. In a table, we can generate something like this,

```
<td>
<div background='white', width={{a}}%></div>
<div background='green', width={{b}}%></div>
<div background='white', width={{c}}%></div>
<td>
```
which hopefully can demonstrate the start time and duration of the step. If width of table is an issue, we can use tooltips to display starting time and duration.

**Comment by BoPeng** [comment]
OK, for workflow

```
[1]
sh:
  echo This is step 1
  sleep 2

[2]
input: for_each={'i': range(3)}
sos_run('b', num = i)

[3]
task:
sh:
  sleep 3
  echo This is step 3

[b_1]
parameter: num=int
output: f'b_{num}'
sh: expand=True
  echo This is b 1
  sleep 4
  touch b_{num}

```
and command
```
$ sos run test -s force -p test.html -d
```

The report looks like:

![image](https://user-images.githubusercontent.com/9889312/41020175-723bea9c-6927-11e8-88bc-6e088bb84a62.png)

**Comment by gaow** [comment]
Looks cool! I just tried on a toy setting my generated statistical benchmark workflow that has layers of nested workflow logic. The outcome looks interesting if you run this command:

```
sos run benchmark.sos -p test.html -d
```

This type of setting might be too complicated anyways, but I guess it helps deciding what makes sense to include in the report.

Test data:

[benchmark.tar.gz](https://github.com/vatlab/SoS/files/2076295/benchmark.tar.gz)

**Comment by BoPeng** [comment]
The benchmark report looks good to me. The only problem is that the filenames are very long so the `input` and `output` columns are very wide. Perhaps we can remove path names from the filenames (but allow showing details when hovering?)

The DAG looks cool although the filenames after step names look wrong. I am not sure if it is a good idea to add output names to nodes anyway. It was introduced to differentiate steps with the same name but different output (e.g. for auxiliary steps) but it is difficult to do because output can be long to display.

**Comment by BoPeng** [comment]
It could be clearer to organize by workflows, basically we can show a line for workflow, and then all steps.  Something like

 ```
workflow
  step
     subworkflow
        subworkflow step
  step
```
But the organization could be difficult if we have nested nested workflows. Perhaps it is easier to do
```
workflow
  step
  step
subworkflow
   subworkflow step
```

Tasks should be separated though because they have different things to display.

Re-organizing steps could help the case with multiple steps with the same names.

**Comment by BoPeng** [comment]
Here is yet another version, with a footer, although perhaps a header makes more sense.

![image](https://user-images.githubusercontent.com/9889312/41067077-e4e6058c-69a9-11e8-8f68-4cac07ee6906.png)

**Comment by BoPeng** [comment]
I do not quite get it. In this particular example you have steps and subworkflows with the same name (e.g. `abs_err`). Do you want to merge subworkflows?

**Comment by gaow** [comment]
>  In this particular example you have steps and subworkflows with the same name (e.g. abs_err)

Indeed, and as shown in the DAG I distinguished them manually with prefix such as `a_`, `b_`. 

If steps or subworkflows of the same name can be put together that might make it easier to read (knowing it is the same computation routine just taking different parameters or inputs), but logically it will change the ordering of the time-line plot and basically make it no more useful than simply report time elapsed. I think this is sort of a dilemma case. Not sure what is the best way out.

**Comment by BoPeng** [comment]
It can help if we change

```
name
```

to
```
id (name)
```
where `id` is the unique ID of the step or subworkflow.

**Comment by BoPeng** [comment]
The new patch allows one to click on the workflow row to toggle step rows. Steps for subworkflows are hidden by default.

![image](https://user-images.githubusercontent.com/9889312/41107332-c5b40310-6a37-11e8-882d-688592d1c2f9.png)

**Comment by gaow** [comment]
It looks better now! 

> where id is the unique ID of the step or subworkflow.

It might help if it does not look ugly. I admit I cannot propose anything satisfactory over this matter. Perhaps `id (name)` is worth a shot.

**Comment by BoPeng** [comment]
Better way to show full filename

![image](https://user-images.githubusercontent.com/9889312/41109460-b35c1dba-6a3c-11e8-8f95-7861f8eb004f.png)

**Comment by gaow** [comment]
yes it looks better!

One question RE `_0` in step name: I see this from time to time. SoS added it to the actual step names. I'm wondering what if users indeed have typed `mean_0` as a step name? My confusion (or worry) is that in the HTML report, users now see some step names that they did not explicitly type out, and get lost there.

**Comment by BoPeng** [comment]
You are right, we also add `default_` to steps such as `[10]`. Need to see if we can make it more consistent.

**Comment by BoPeng** [comment]
I have stopped adding `_0` to step name (`default` is kept) but I am not sure if this would break anything. Let us see.

**Comment by gaow** [comment]
One minor issue maybe -- the gif file name for dag is `sosDot.gif`. Should we use the SoS file name instead in case multiple SoS runs are done under the same directory?

**Comment by BoPeng** [comment]
I have added file size to file list because this is important information for my pipelines and a good way to demonstrate the scale of the analysis and sometimes if a step is completed successful (size 0).

The gif problem is resolved. I have also removed `[]` from node name if there is no input.

**Comment by BoPeng** [comment]
I think I am more or less done here.

The only feature that I kind of like is to show "xxx days ago" instead of "YYMMDD", which makes it clear when the report was generated. ..

**Comment by gaow** [comment]
Very nice! I'll see how it plays with my larger benchmarks. 

It would be real nice to be able to hover over "png" or "jpg" suffixed files in the drop-down menu and get a pop up figure -- you can imaging it be useful in quick diagnostics if the diagnostics are made possible in the form of figures within a step. Just a thought :)

**Comment by BoPeng** [comment]
Browser cannot reach local files so you cannot preview files from a webpage. The only way to proceed would be to include/embed the figures in the report (as thumbnails) but that means the report file can be huge....

**Comment by gaow** [comment]
I see. Never mind ... that is not feasible.
